File: C++/Problems/Problems/problem_898.cpp
#include "problem_898.h"

#include <unordered_set>

bool problem_898::test() {
    std::vector arr { 1, 2, 4 };

    auto expected = 6;

    auto result = subarrayBitwiseORs(arr);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_898::subarrayBitwiseORs(std::vector<int> &arr) {
    std::unordered_set<int> result;
    std::unordered_set partial{0};
    for (auto num : arr)
    {
        std::unordered_set<int> tmp;
        for (auto prev : partial)
            tmp.insert(num | prev);
        tmp.insert(num);
        for (auto value : tmp)
            result.insert(value);
        partial = tmp;
    }

    return result.size();
}

File: Python/Problems/Problems/problem898.py
import math
from typing import List, Dict

from .problem import Problem


class Problem898(Problem):
    def test(self) -> bool:
        arr = [1, 2, 4]

        expected = 6

        result = self.subarrayBitwiseORs(arr)

        print(result)

        return result == expected

    def subarrayBitwiseORs(self, arr: List[int]) -> int:
        result = set()
        partial = {0}
        for num in arr:
            tmp = set()
            for prev in partial:
                tmp.add(num | prev)
            tmp.add(num)
            for value in tmp:
                result.add(value)
            partial = tmp

        return len(result)